keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1076 @@
+3. The user watched the status of PodGroup. Pods associated with this PodGroup are expected to be co-scheduled if resources are adequate.
+4. The user waits for the PodGroup to be complete and then deletes it.
+
+### Scheduler
+#### QueueSort
+In order to maximize the chance that the pods which belong to the same `PodGroup` to be scheduled consecutively, we need to implement a customized `QueueSort` plugin to sort the Pods properly.
+
+Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.
+
+Secondly, if two Pods hold the same priority, the sorting precedence is described as below:
+
+- If they are both regularPods (without pod.spec.podGroup), compare their `InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned ahead of the other.
+  
+- If one is regularPod and the other is pgPod (without pod.spec.podGroup), compare regularPod's `InitialAttemptTimestamp` with the `creationtime` of pgPod's PodGroup: the Pod with earlier timestamp is positioned ahead of the other.
+  
+- If they are both pgPods:
+  - Compare their pod group's `CreationTimestamp`: the Pod in pod group which has earlier timestamp is positioned ahead of the other.
+  - If their `CreationTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `CreationTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)
+
+#### PreFilter
+In PreFilter phase, we need to check if pod.spec.podGroup is nil firstly. If so, return success. If not, check if the pod group is existed in the scheduler internal cache and pod.spec.podGroup.subset is matched in pod group, if not, reject the pod in PreFilter.
+
+
+#### Permit
+In Permit phase, It is divided into two parts.
+
+- The first part is that we put the pods that doesn't meet MinMember into the WaitingMap and reserve resources until MinMember are met or timeout is triggered.
+
+   - Get the number of Running pods that belong to the same PodGroup
+   - Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
+
+  If pods in all the subsets meet `Running + WaitingPods + 1 >= MinMember(1 means the pod itself) `, approve the waiting pods that belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (the timeout comes from the ScheduleTimeoutSeconds in PodGroup.).
+
+- Then second part is in order for pods belonging to the same PodGroup to be scheduled as a whole without being broken up, we will leverage the feature added https://github.com/kubernetes/kubernetes/pull/103383 to move pods of the same PodGroup back to internal scheduling activeQ instantly.
+
+#### UnReserve
```

Comment:
WaitForFirstConsumer is happened in the pre-bind phase., after the premit phase. Won't be affected by gang scheduling here.

Reply:
One limitation of the delayed volume binding is that it currently doesn't have the notion that a group of volumes would want to be provisioned together. So what could happen is you have a volume provisioned successfully for one pod, but then fails for the second. Or even if you have multiple volumes in a single pod. 

It would be interesting to explore if this could help address that use case too.